cmd/erasure-healfile.go

```diff
@@ -126,27 +127,48 @@ func (s ErasureStorage) HealFile(staleDisks []StorageAPI, volume, path string,
 
 		// write computed shards as chunks on file in each
 		// stale disk
+		writeSucceeded := false
 		for i, disk := range staleDisks {
-			if disk == nil {
+			// skip nil disk or disk that had error on
+			// previous write
+			if disk == nil || writeErrors[i] != nil {
 				continue
 			}
 
-			err = disk.AppendFile(healVolume, healPath, blocks[i])
-			if err != nil {
-				return f, traceError(err)
+			writeErrors[i] = disk.AppendFile(dstVol, dstPath, blocks[i])
+			if writeErrors[i] == nil {
+				hashers[i].Write(blocks[i])
+				writeSucceeded = true
 			}
-			hashers[i].Write(blocks[i])
+		}
+
+		// If all disks had write errors we quit.
+		if !writeSucceeded {
+			// build error from all write errors
+			return f, traceError(joinWriteErrors(writeErrors))
 		}
 	}
 
 	// copy computed file hashes into output variable
 	f.Size = size
-	f.Algorithm = algorithm
+	f.Algorithm = alg
 	for i, disk := range staleDisks {
-		if disk == nil {
+		if disk == nil || writeErrors[i] != nil {
 			continue
 		}
 		f.Checksums[i] = hashers[i].Sum(nil)
 	}
 	return f, nil
 }
+
+func joinWriteErrors(errs []error) error {
+	msgs := []string{}
+	for i, err := range errs {
+		if err == nil {
+			continue
```

Comment:
This would not be correct. `errs[i] == nil` if `staleDisks[i] == nil` (i.e. a non-stale disk). To check if all stale disks had write errors, it would be required to pass the `staleDisks` slice to `joinWriteErrors()` and that seems more complex than using `writeSucceeded` to check if at least one disk was written.

Reply:
I see, agreed. Just tried to avoid state carrying within the function but seems to be easiest solution...